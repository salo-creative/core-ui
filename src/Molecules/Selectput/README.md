# Selectput

An opinionated way to select from a list and then input a value against that list.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Selectput from '@salo/core-ui/Molecules/Selectput';
```

Implement as follows

```javascript
const [value, setValue] = React.useState('1');

<Selectput 
  options={ [
    { value: '1', label: 'Option 1' },
    { value: '2', label: 'Option 2' },
    { value: '3', label: 'Option 3' }
  ] }
  value={ value }
  onSubmit={ (val) => setValue(val) }
/>
```