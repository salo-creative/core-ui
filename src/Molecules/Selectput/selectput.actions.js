import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';

const Actions = ({ disabled, handleSubmit, handleClose }) => (
  <React.Fragment>
    <button
      type='button'
      onClick={ handleSubmit }
      disabled={ disabled.submit }
      className='salo-selectput__button--submit'
    >
      <Icon icon='tick' size={ 16 } />
    </button>
    <button
      type='button'
      onClick={ handleClose }
      disabled={ disabled.close }
      className='salo-selectput__button--close'
    >
      <Icon icon='close' size={ 16 } />
    </button>
  </React.Fragment>
);

Actions.defaultProps = {
  disabled: {
    submit: false,
    close: false
  }
};

Actions.propTypes = {
  disabled: PropTypes.shape({
    submit: PropTypes.bool.isRequired,
    close: PropTypes.bool.isRequired
  }),
  handleClose: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired
};

export default Actions;