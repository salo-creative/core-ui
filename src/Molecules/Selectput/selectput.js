import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';

import Actions from './selectput.actions';

const Selectput = ({
  className,
  options,
  onChange,
  onReset,
  onSubmit,
  placeholder,
  initialMode,
  initialSelected,
  initialValue,
  renderItem,
  icons
}) => {
  const [mode, setMode] = React.useState(initialMode);
  const [selected, setSelected] = React.useState(initialSelected);
  const [value, setValue] = React.useState(initialValue);
  const inputEl = React.useRef(null);

  const handleSelect = (e) => {
    if (typeof onChange === 'function') {
      onChange(e);
    }
    setSelected(e.target.value);
    setMode('edit');
  };
  
  const handleSubmit = (event) => {
    onSubmit({
      type: selected,
      value,
      event
    });
   
    setMode('default');
  };

  const handleChange = (e) => {
    if (typeof onChange === 'function') {
      onChange(e);
    }
    setValue(e.target.value);
  };

  const handleClose = (e) => {
    if (typeof onReset === 'function') {
      onReset(e);
    }
    setMode('default');
  };

  React.useEffect(() => {
    // Autofocus element when editing.
    if (mode === 'edit') {
      inputEl.current.focus();
    }
  }, [mode]);

  switch (mode) {
    case 'edit':
      return (
        <div className={ `${ className } salo-selectput salo-selectput--${ mode }` }>
          <input
            className='salo-selectput__input'
            onChange={ handleChange }
            onKeyPress={ (event) => {
              if (event.key === 'Enter') {
                handleSubmit(event);
              }
            } }
            ref={ inputEl }
          />
          { icons && (
            <span className='salo-selectput__icon-wrapper'>
              <Icon icon={ icons[selected] || selected } fill='#fff' size={ 16 } />
            </span>
          ) }
          <Actions
            handleSubmit={ handleSubmit }
            handleClose={ handleClose }
          />
        </div>
      );
    case 'select':
      return (
        <div className={ `${ className } salo-selectput salo-selectput--${ mode }` }>
          <div className='salo-selectput__button'>
            { placeholder }
            <div className='salo-selectput__wrapper'>
              <ul
                className='salo-selectput__list'
              >
                { options.map((item, index) => {
                  if (typeof renderItem === 'function') {
                    return renderItem({
                      item, index, handleSelect
                    });
                  }
                  return (
                    <li key={ item.value }>
                      <button
                        className='salo-selectput__item'
                        type='button'
                        onClick={ handleSelect }
                        value={ item.value }
                      >
                        { item.label }
                      </button>
                    </li>
                  );
                }) }
              </ul>
            </div>
          </div>
          <Actions
            disabled={ {
              submit: true, close: false
            } }
            handleSubmit={ handleSubmit }
            handleClose={ handleClose }
          />
        </div>
      );
    default:
      return (
        <div className={ `${ className } salo-selectput salo-selectput--${ mode }` }>
          <button
            type='button'
            className='salo-selectput__button'
            onFocus={ () => setMode('select') }
            onClick={ () => setMode('select') }
          >
            { placeholder }
          </button>
          <Actions
            disabled={ {
              submit: true, close: false
            } }
            handleSubmit={ handleSubmit }
            handleClose={ handleClose }
          />
        </div>
      );
  }
};

Selectput.defaultProps = {
  className: null,
  icons: null,
  initialMode: 'default',
  initialValue: '',
  initialSelected: null,
  onChange: null,
  onReset: null,
  placeholder: 'Please selectâ€¦',
  renderItem: null
};

Selectput.propTypes = {
  className: PropTypes.string,
  icons: PropTypes.object,
  initialMode: PropTypes.string,
  initialSelected: PropTypes.string,
  initialValue: PropTypes.string,
  onChange: PropTypes.func,
  onReset: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.any.isRequired
  })).isRequired,
  placeholder: PropTypes.string,
  renderItem: PropTypes.func
};

export default Selectput;