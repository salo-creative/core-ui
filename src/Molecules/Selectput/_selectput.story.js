import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Selectput } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Molecules | Selectput', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['selector'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    return (
      <RenderWithProps store={ store }>
        <Selectput
          options={ [
            { value: 'linkedin', label: 'LinkedIn' },
            { value: 'phone', label: 'Phone' },
            { value: 'skype', label: 'Skype' }
          ] }
          value={ store.value }
          onChange={ (value) => store.set({ value }) }
          onReset={ () => store.set({ value: '' }) }
          onSubmit={ (value) => store.set({ value }) }
        />
      </RenderWithProps>
    );
  }),
  { info: { propTablesExclude: [] }, notes: README }
);


stories.add(
  'Edit',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    return (
      <RenderWithProps store={ store }>
        <Selectput
          options={ [
            { value: 'linkedin', label: 'LinkedIn' },
            { value: 'phone', label: 'Phone' },
            { value: 'skype', label: 'Skype' }
          ] }
          value={ store.value }
          initialMode='edit'
          initialSelected='phone'
          onChange={ (value) => store.set({ value }) }
          onReset={ () => store.set({ value: '' }) }
          onSubmit={ (value) => store.set({ value }) }
        />
      </RenderWithProps>
    );
  }),
  { info: { propTablesExclude: [] }, notes: README }
);