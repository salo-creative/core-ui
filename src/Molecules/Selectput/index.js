import Selectput from './selectput';

export default Selectput;