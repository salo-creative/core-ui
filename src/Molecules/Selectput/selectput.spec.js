import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Selectput from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Selectput />
  );
  expect(container.firstChild).toMatchSnapshot();
});