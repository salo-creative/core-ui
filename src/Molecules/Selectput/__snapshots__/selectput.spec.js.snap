// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  height: 16px;
  width: 16px;
  margin: 0;
}

<div
  class="null salo-selectput salo-selectput--default"
>
  <button
    class="salo-selectput__button"
    type="button"
  >
    Please selectâ€¦
  </button>
  <button
    class="salo-selectput__button--submit"
    disabled=""
    type="button"
    value="Add"
  >
    <svg
      aria-hidden="true"
      class="c0"
      focusable="false"
      size="16"
      viewBox="0 0 24 24"
    >
      <path
        d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
        fill="#00222b"
      />
    </svg>
  </button>
  <button
    class="salo-selectput__button--close"
    type="button"
    value="Close"
  >
    <svg
      aria-hidden="true"
      class="c0"
      focusable="false"
      size="16"
      viewBox="0 0 24 24"
    >
      <path
        d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
        fill="#00222b"
      />
    </svg>
  </button>
</div>
`;
